src/ripple/app/ledger/impl/DisputedTx.h

```diff
@@ -73,26 +77,158 @@ class DisputedTx
-    Serializer transaction;
+    Tx_t transaction;
 
-    hash_map <NodeID, bool> mVotes;
+    hash_map <NodeID_t, bool> mVotes;
     beast::Journal j_;
 };
 
+// Track a peer's yes/no vote on a particular disputed transaction
+template <class Traits>
+void DisputedTx<Traits>::setVote (NodeID_t const& peer, bool votesYes)
+{
+    auto res = mVotes.insert (std::make_pair (peer, votesYes));
+
+    // new vote
+    if (res.second)
+    {
+        if (votesYes)
+        {
+            JLOG (j_.debug())
+                    << "Peer " << peer << " votes YES on " << mTransactionID;
+            ++mYays;
+        }
+        else
+        {
+            JLOG (j_.debug())
+                    << "Peer " << peer << " votes NO on " << mTransactionID;
+            ++mNays;
+        }
+    }
+    // changes vote to yes
+    else if (votesYes && !res.first->second)
+    {
+        JLOG (j_.debug())
+                << "Peer " << peer << " now votes YES on " << mTransactionID;
+        --mNays;
+        ++mYays;
+        res.first->second = true;
+    }
+    // changes vote to no
+    else if (!votesYes && res.first->second)
+    {
+        JLOG (j_.debug())
+                << "Peer " << peer << " now votes NO on " << mTransactionID;
+        ++mNays;
+        --mYays;
+        res.first->second = false;
+    }
+}
+
+// Remove a peer's vote on this disputed transasction
+template <class Traits>
+void DisputedTx<Traits>::unVote (NodeID_t const& peer)
+{
+    auto it = mVotes.find (peer);
+
+    if (it != mVotes.end ())
+    {
+        if (it->second)
+            --mYays;
+        else
+            --mNays;
+
+        mVotes.erase (it);
+    }
+}
+
+template <class Traits>
+bool DisputedTx<Traits>::updateVote (int percentTime, bool proposing)
+{
+    if (mOurVote && (mNays == 0))
+        return false;
+
+    if (!mOurVote && (mYays == 0))
+        return false;
+
+    bool newPosition;
+    int weight;
+
+    if (proposing) // give ourselves full weight
+    {
+        // This is basically the percentage of nodes voting 'yes' (including us)
+        weight = (mYays * 100 + (mOurVote ? 100 : 0)) / (mNays + mYays + 1);
+
+        // VFALCO TODO Rename these macros and turn them into language
+        //             constructs.  consolidate them into a class that collects
+        //             all these related values.
+        //
+        // To prevent avalanche stalls, we increase the needed weight slightly
+        // over time.
+        if (percentTime < AV_MID_CONSENSUS_TIME)
+            newPosition = weight >  AV_INIT_CONSENSUS_PCT;
+        else if (percentTime < AV_LATE_CONSENSUS_TIME)
+            newPosition = weight > AV_MID_CONSENSUS_PCT;
+        else if (percentTime < AV_STUCK_CONSENSUS_TIME)
+            newPosition = weight > AV_LATE_CONSENSUS_PCT;
+        else
+            newPosition = weight > AV_STUCK_CONSENSUS_PCT;
+    }
+    else
+    {
+        // don't let us outweigh a proposing node, just recognize consensus
```

Comment:
I tried to do this. It doesn't work. It needs to be in scope for the following if. And that's the entire function at that point, so there's nothing left to exclude it from.